File: DesignPatterns/Creational/Builder/PizzaClient.java
//Product
class Pizza {
  private String dough;
  private String sauce;
  private String topping;

  public String getDough(){
    return this.dough;
  }
  public void setDough(String dough) {
    this.dough = dough;
  }
  public String getSauce(){
    return this.sauce;
  }
  public void setSauce(String sauce) {
    this.sauce = sauce;
  }
  public String getTopping(){
    return this.topping;
  }
  public void setTopping(String topping) {
    this.topping = topping;
  }

  public String toString() {
    return "Pizza with : " + this.dough + " * " + this.sauce + " * " +  this.topping;
  }
}

//Abstract builder
abstract class PizzaBuilder {

  protected Pizza pizza;

  public Pizza getPizza(){
    return pizza;
  }

  public void createNewPizza() {
    pizza = new Pizza();
  }

  public abstract void buildDough();
  public abstract void buildSauce();
  public abstract void buildTopping();
}

//Concrete builder
class HawaiianPizzaBuilder extends PizzaBuilder {

  public void buildDough(){
    pizza.setDough("cross");
  }

  public void buildSauce(){
    pizza.setSauce("mild");
  }

  public void buildTopping(){
    pizza.setTopping("ham+pineapple");
  }
}

//Another concrete builder
class SpicyPizzaBuilder extends PizzaBuilder {

  public void buildDough() {
    pizza.setDough("pan baked");
  }

  public void buildSauce(){
    pizza.setSauce("hot");
  }

  public void buildTopping(){
    pizza.setTopping("pepperoni+salami");
  }
}

//Director
class Waiter {

  private PizzaBuilder pizzaBuilder;

  public Waiter(PizzaBuilder pizzaBuilder){
    this.pizzaBuilder = pizzaBuilder;
  }

  public Pizza getPizza(){
    return pizzaBuilder.getPizza();
  }

  public void buildPizza(){
    pizzaBuilder.createNewPizza();
    pizzaBuilder.buildDough();
    pizzaBuilder.buildSauce();
    pizzaBuilder.buildTopping();
  }
}

//Client
public class PizzaClient {
  public static void main(String[] args) {
    HawaiianPizzaBuilder hp = new HawaiianPizzaBuilder();
    SpicyPizzaBuilder sp = new SpicyPizzaBuilder();
    Waiter wt = new Waiter(hp);
    Waiter wt1 = new Waiter(sp);
    wt.buildPizza();
    wt1.buildPizza();
    System.out.println(wt.getPizza());
    System.out.println(wt1.getPizza());
  }
}

File: DesignPatterns/Creational/Builder/Ruby/pizza_client.rb
class Pizza
  attr_accessor :dough, :sauce, :topping

  def to_s
    "Pizza with : #{self.dough} * #{self.sauce} * #{self.topping}"
  end
end

class PizzaBuilder
  attr_accessor :pizza
  
  def initialize
    self.pizza = Pizza.new
  end

  def build_dough
    raise NotImplementedError
  end

  def build_sauce
    raise NotImplementedError
  end

  def build_topping
    raise NotImplementedError
  end

  def fetch_pizza
    return self.pizza
  end
end

class HawaiianPizzaBuilder < PizzaBuilder
  def build_dough
    pizza.dough = 'cross'
  end

  def build_sauce
    pizza.sauce = 'mild'
  end

  def build_topping
    pizza.topping = 'ham+pineapple'
  end
end

class SpicyPizzaBuilder < PizzaBuilder
  def build_dough
    pizza.dough = 'pan baked'
  end

  def build_sauce
    pizza.sauce = 'hot'
  end

  def build_topping
    pizza.topping = 'pepperoni+salami'
  end
end

class Waiter
  attr_accessor :pizza_builder

  def initialize(pizza_builder)
    self.pizza_builder = pizza_builder
  end

  def build_pizza
    self.pizza_builder.build_dough
    self.pizza_builder.build_sauce
    self.pizza_builder.build_topping
  end

  def fetch_pizza
    self.pizza_builder.fetch_pizza
  end
end

hp = HawaiianPizzaBuilder.new
sp = SpicyPizzaBuilder.new
w1 = Waiter.new(hp);
w2 = Waiter.new(sp);

w1.build_pizza
w2.build_pizza

puts w1.fetch_pizza
puts w2.fetch_pizza